cljfmt/src/cljfmt/main.clj

```diff
@@ -154,71 +154,88 @@
       (read-string contents)
       (edn/read-string contents))))
 
-(def default-options
+(defn- find-config-file []
+  (loop [dir (.getAbsoluteFile (java.io.File. ""))]
+    (when dir
+      (let [file1 (java.io.File. dir ".cljfmt.edn")
+            file2 (java.io.File. dir ".cljfmt.clj")]
+        (cond
+          (.exists file1) (.getAbsolutePath file1)
+          (.exists file2) (.getAbsolutePath file2)
+          :else (recur (.getParentFile dir)))))))
+
+(defn- load-configuration []
+  (some-> (find-config-file)
+          cli-file-reader))
```

Comment:
This would be better written as:

```clojure
(defn find-config-file [filename]
  (loop [dir (.getAbsoluteFile. (io/file "."))]
    (when dir
      (let [file (io/file dir filename)]
        (if (.exists file)
          file
          (recur (.getParentFile dir)))))))

(defn- load-configuration []
  (some-> (or (find-config-file ".cljfmt.edn")
              (find-config-file ".cljfmt.clj"))
          cli-file-reader))
```

Reply:
I considered that, but that'd find and use a `.cljfmt.edn` in `$HOME` (for instance) and ignore a closer `.cljfmt.clj` in the project directory.
I also considered passing both as arguments, but I think that makes `find-config-file` much more complicated and probably is not something that needs to be dynamic.

Reply:
Excellent point. In which case, perhaps:

```clojure
(defn- find-config-file [pred]
  (loop [dir (.getAbsoluteFile. (io/file "."))]
    (when dir
      (or (first (filter pred (.listFiles dir)))
          (recur (.getParentFile dir))))))

(def ^:private config-filenames #{".cljfmt.edn" ".cljfmt.clj"})

(defn- load-configuration []
  (some-> (find-config-file #(config-filenames (.getName %)))
          cli-file-reader))

Reply:
That's similar to my approach in passing the filenames as arguments, I also considered a loop over the filenames, but I still think both are more complicated without a real need.
It also seems inefficient, as it now needs to do a directory listing in every parent directory, which could be large or there might be permission issues. In particular when used as a formatter with an editor, this sort of thing can impact start-up time noticeably.

Since these filenames are pretty static, I think we should test them explicitly.

Reply:
I generally don't like talking about efficiency without benchmarks, as it may be the case that even very large directories take little time to search through.

However, the permissions issues argument is a very good one. The process may have access to the file without having access to list the directory. So what about:

```clojure
(defn- parent-dirs [root]
  (->> (.getAbsoluteFile (io/file root))
       (iterate #(.getParentFile %))
       (take-while some?)))

(defn- find-file-in-dir [dir name]
  (let [f (io/file dir name)]
    (when (.exists f) f)))

(defn- find-config-file-in-dir [dir]
  (or (find-file-in-dir dir ".cljfmt.edn")
      (find-file-in-dir dir ".cljfmt.clj")

(defn- load-configuration []
  (some->> (parent-dirs ".")
           (some find-config-file-in-dir)
           cli-file-reader))
```

I believe your original design had a similar `parent-dirs` function?

Reply:
Similar, yes. I used your version, added type hinting for GraalVM. The file also has to be converted to a string before `cli-file-reader` because there's an extension check there, which is a bit ugly, perhaps you have a better idea.
I think this likely was a problem in my previous iteration already. The branch I use personally doesn't have the `.clj` reading because it has a big performance impact.